File: Week2/공통/미로탈출/suhyun1.java
package Week2.공통.미로탈출;

//7:40 ~ 8:00

import java.util.LinkedList;
import java.util.Queue;

public class suhyun1 {
    int sx = 0, sy = 0, lx = 0, ly = 0, ex = 0, ey = 0;
    int[][] dir = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    class Node{
        int x = 0, y = 0, cnt = 0;

        public Node(int x, int y, int cnt){
            this.x = x;
            this.y = y;
            this.cnt = cnt;
        }
    }

    public int bfs(String[] maps, int startX, int startY, int targetX, int targetY){
        boolean[][] visited = new boolean[maps.length][maps[0].length()];
        Queue<Node> q = new LinkedList<Node>();
        q.offer(new Node(startX, startY, 0));
        visited[startX][startY] = true;
        while(!q.isEmpty()){
            Node n = q.poll();
            if(n.x == targetX && n.y == targetY){
                return n.cnt;
            }

            for(int i=0; i<4; i++){
                int nx = n.x + dir[i][0];
                int ny = n.y + dir[i][1];

                if(nx < 0 || ny < 0 || nx >= maps.length || ny >= maps[0].length()){
                    continue;
                }

                if(maps[nx].charAt(ny) == 'X'){
                    continue;
                }

                if(!visited[nx][ny]){
                    q.offer(new Node(nx, ny, n.cnt + 1));
                    visited[nx][ny] = true;
                }
            }
        }
        return 0;
    }

    public int solution(String[] maps) {
        for(int i=0; i<maps.length; i++){
            for(int j=0; j<maps[i].length(); j++){
                if(maps[i].charAt(j) == 'S'){
                    sx = i;
                    sy = j;
                } else if (maps[i].charAt(j) == 'L'){
                    lx = i;
                    ly = j;
                } else if(maps[i].charAt(j) == 'E'){
                    ex = i;
                    ey = j;
                }
            }
        }

        int before = bfs(maps, sx, sy, lx, ly);
        int after = bfs(maps, lx, ly, ex, ey);

        if (before == 0 || after == 0){
            return -1;
        }

        return before + after;
    }
}

File: Week2/공통/미로탈출/suhyun1.py
# queue = deque()
# while queue:
# nx, ny, cnt = queue.popleft()
from collections import deque

sx, sy, lx, ly, ex, ey = 0, 0, 0, 0, 0, 0
dir = [[0,1], [1, 0], [0, -1], [-1, 0]]

def bfs(maps, startX, startY, tX, tY):
    visited = [[False] * len(maps[0]) for _ in range(len(maps))]
    queue = deque()
    queue.append([startX, startY, 0])
    visited[startX][startY] = True
    while queue:
        x, y, cnt = queue.popleft()
        if x == tX and y == tY:
            return cnt
        
        for dx, dy in dir:
            nx = x + dx
            ny = y + dy
            
            if nx < 0 or ny < 0 or nx >= len(maps) or ny >= len(maps[0]):
                continue
            
            if maps[nx][ny] == 'X':
                continue
            
            if not visited[nx][ny]:
                queue.append([nx, ny, cnt + 1])
                visited[nx][ny] = True
    return 0       

def solution(maps):
    
    for i, row in enumerate(maps):
        for j, col in enumerate(row):
            if col == 'S':
                sx, sy = i, j
            elif col == 'E':
                ex, ey = i, j
            elif col == 'L':
                lx, ly = i, j
    
    before = bfs(maps, sx, sy, lx, ly)
    after = bfs(maps, lx, ly, ex, ey)
    if before == 0 or after == 0:
        return -1
            
    return before + after

print(solution(["SOOOL","XXXXO","OOOOO","OXXXX","OOOOE"]))